src/platform/ESP32/MdnsImpl.cpp

```diff
@@ -0,0 +1,241 @@
+        current = results;
+
+        for (size_t i = 0; i < serviceSize; i++, current = current->next)
+        {
+            mdns_ip_addr_t * addr          = current->addr;
+            mdns_ip_addr_t * preferredAddr = addr;
+
+            while (addr != nullptr)
+            {
+                if (addr->addr.type == ESP_IPADDR_TYPE_V6)
+                {
+                    preferredAddr = addr;
+
+                    break;
+                }
+                addr = addr->next;
+            }
+            if (preferredAddr && preferredAddr->addr.type == ESP_IPADDR_TYPE_V6)
+            {
+                ip6_addr_t addrBinary;
+
+                memcpy(&addrBinary, &preferredAddr->addr.u_addr.ip6, sizeof(addrBinary));
+                services[i].mAddress.SetValue(chip::Inet::IPAddress::FromIPv6(addrBinary));
+            }
+            else if (preferredAddr && preferredAddr->addr.type == ESP_IPADDR_TYPE_V4)
+            {
+                ip4_addr_t addrBinary;
+
+                memcpy(&addrBinary, &preferredAddr->addr.u_addr.ip4, sizeof(addrBinary));
+                services[i].mAddress.SetValue(chip::Inet::IPAddress::FromIPv4(addrBinary));
+            }
+            if (current->instance_name)
+            {
+                strncpy(services[i].mName, current->instance_name, sizeof(services[i].mName));
+            }
+            services[i].mPort     = current->port;
+            services[i].mProtocol = protocol;
+            strncpy(services[i].mType, type, sizeof(services[i].mType));
+        }
+        ChipLogProgress(DeviceLayer, "Call callback");
+        callback(context, services, serviceSize, CHIP_NO_ERROR);
+    }
+
+exit:
+    if (services != nullptr)
+    {
+        chip::Platform::MemoryFree(services);
+    }
+    if (results != nullptr)
+    {
+        mdns_query_results_free(results);
+    }
+
+    return error;
+}
+
+CHIP_ERROR ChipMdnsResolve(MdnsService * service, chip::Inet::InterfaceId interface, MdnsResolveCallback callback, void * context)
+{
+    CHIP_ERROR error           = CHIP_NO_ERROR;
+    mdns_result_t * textResult = nullptr;
+    TextEntry * textEntries;
+    esp_err_t espError;
+
+    VerifyOrExit(service != nullptr, error = CHIP_ERROR_INVALID_ARGUMENT);
+    espError = mdns_query_txt(service->mName, service->mType, GetProtocolString(service->mProtocol), kTimeoutMilli, &textResult);
+    VerifyOrExit(espError == ESP_OK && textResult != nullptr, error = CHIP_ERROR_INTERNAL);
+
+    if (textResult->txt_count > 0)
+    {
+        textEntries = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(textResult->txt_count, sizeof(TextEntry)));
+        for (size_t i = 0; i < textResult->txt_count; i++)
+        {
+            textEntries[i].mKey      = textResult->txt[i].key;
+            textEntries[i].mData     = reinterpret_cast<const uint8_t *>(textResult->txt[i].value);
+            textEntries[i].mDataSize = strlen(textResult->txt[i].value);
```

Comment:
Need to use strnlen here, and cap size to the max CHIP-specified TXT field (which is like < 16 bytes)